Documentation update for job type development (#223)
TheRacetrack/racetrack: docs/development/plugins-job-types.md

    }
}
```
</details>

`go_wrapper/openapi.go` handles the swaggerui dashboard:
<details>
  <summary>File `go_wrapper/openapi.go`</summary>

```go
package main

import (
    "net/http"
    "os"
    "text/template"

    "github.com/gin-gonic/gin"
)

func MountOpenApi(router *gin.Engine, baseUrl string) {
    jobName := os.Getenv("JOB_NAME")
    jobVersion := os.Getenv("JOB_VERSION")

    router.Any(baseUrl+"/static/openapi.json", func(c *gin.Context) {
        tmplt := template.New("openapi.json")
        tmplt, _ = tmplt.ParseFiles("./swaggerui/openapi.json")

        context := map[string]string{
            "jobName":    jobName,
            "jobVersion": jobVersion,
        }
        tmplt.Execute(c.Writer, context)
    })
    router.Any(baseUrl+"/static/:filename", func(c *gin.Context) {
        filename := c.Param("filename")
        http.ServeFile(c.Writer, c.Request, "./swaggerui/"+filename)
    })

    router.Any(baseUrl+"/", func(c *gin.Context) {
        http.ServeFile(c.Writer, c.Request, "./swaggerui/index.html")
    })
}
```
</details>

`go_wrapper/go.mod` and `go_wrapper/go.sum` are the files with package dependencies used by Go:
<details>

>   <summary>File `go_wrapper/go.mod`</summary>

Reviewer: Maybe we could point to the link https://github.com/TheRacetrack/plugin-go-job-type/blob/master/golang-job-type/go_wrapper/go.sum instead of copying this huge file here, which makes it also harder to maintain. 

Same with the `metrics.go` and `openapi.go` files which are non-obligatory for making something a job type.
I'm afraid that too complex job type documentation might scare off the developers.

# Creating Job Type plugins

This document covers how to create your own [job type](../glossary.md) plugin
to extend Racetrack.

## Quickstart Step-by-step
1. Create a git repository for your plugin

2. Create a [plugin manifest](../manifest-schema.md) in a plugin subdirectory

> 3. Write a wrapper for the software you are making a job type for

Reviewer: Nitpick: A dot improves reading here
```suggestion
3. Write a wrapper for the software you are making a job type for.
```

- You MAY expose any other endpoints.
- Be careful to isolate libraries / requirements installed by the user
  from the versions of the libraries used by the core wrapper.

## Example job type

The [Go job type](https://github.com/TheRacetrack/plugin-go-job-type) is a
fully featured job type maintained by the racetrack team that serves as an
example job type that implements all features (including optional ones)
provided by racetrack. A barebones quickstart version of said jobtype
following the guide above would look as follows:

### 1. Create a git repository

Create https://github.com/TheRacetrack/plugin-go-job-type.

### 2. Create a plugin manifest in a plugin subdirectory

Create `golang-job-type` subdirectory, create `plugin-manifest.yaml` in it.
It should look as follows:

```yaml
name: golang-job-type
version: 1.3.0
url: https://github.com/TheRacetrack/plugin-go-job-type
```

### 3. Write a wrapper for running Go code

Create the `go_wrapper` subdirectory in the `golang-job-type` subdirectory.
It should look like:
```
go_wrapper
├── go.mod
├── go.sum
├── handler
│   ├── go.mod
│   ├── go.sum
│   └── perform.go
├── health.go
├── main.go
├── Makefile
└── server.go
```

> `go_wrapper/src/handler/` is for handling the user's code, it will be

Reviewer: ```suggestion
`go_wrapper/handler/` is for handling the user's code, it will be
```

injected there by docker when building the image. It looks like this:

```go
// This is just a stub for IDE.

> // It gets replaced by user's Job code in wrappers/docker/golang/job-template.Dockerfile

Reviewer: This comment is a bit out-of-date
```suggestion
// It gets replaced by user's Job code in golang-job-type/job-template.Dockerfile
```

RUN chmod -R a+rw /src/go_wrapper && cd /src/go_wrapper/ && go mod download
# Build Go Job
RUN go get ./... && go build -o go_wrapper
# Set environment variables that are expected by Job executable
ENV JOB_NAME "{{ manifest.name }}"
ENV JOB_VERSION "{{ manifest.version }}"
ENV GIT_VERSION "{{ git_version }}"
ENV DEPLOYED_BY_RACETRACK_VERSION "{{ deployed_by_racetrack_version }}"
```
</details>

<details>
  <summary>File `go-job-type/plugin.py`</summary>

```python
from __future__ import annotations
from pathlib import Path
class Plugin:
    def job_types(self) -> dict[str, tuple[Path, Path]]:
        """
        Job types provided by this plugin
        :return dict of job type name (with version) -> (base image path, dockerfile template path)
        """
        return {
            f'golang:{self.plugin_manifest.version}': (
                self.plugin_dir / 'base.Dockerfile',
                self.plugin_dir / 'job-template.Dockerfile',
            ),
        }
```
</details>

Finally, also create the `.racetrackignore` file:
```
Makefile
go.sum
```

### 8. Bundle plugin into a ZIP file
Install the racetrack client
```shell
python3 -m pip install --upgrade racetrack-client
```

> And then run `racetrack plugin bundle` in the `go-job-type` directory.

Reviewer: How about adding a "Frequently Asked Questions" section at the bottom?
It may explain a difference between base image and job template, which in my opinion can be a source of a common confusion.

## Frequently Asked Questions
### What distinguishes a base image from a job template?

The job image creation process involves two stages:

1. **Preparing the base image** - 
  The base image comprises files that are common to all jobs, such as your wrapper code, and isn't tied to any specific job or manifest.

2. **Building a job from the template** - 
  Before a job is deployed, the job template is combined with a user-submitted manifest to create the final Dockerfile that will be used to build the job image.
  The job image is an extension of the base image, with the job source code added during the building process. This stage is unique to each job.

Author: I just created an FAQ section, is it okay or would you rather I just took what you've written here?

Reviewer: It's alright.
